File: packages/agents/src/newsapi-fetcher.ts
/**
 * NewsAPI Fetcher Agent
 *
 * Reads active NewsAPI sources from the DB (kind='newsapi', url=query string),
 * hits the NewsAPI /everything endpoint, and inserts new articles as drafts.
 *
 * Required env: DATABASE_URL, NEWSAPI_KEY
 * Run: pnpm --filter @fluxnews/agents newsapi
 */

import { db, sources, posts } from '@fluxnews/db';
import { eq, and } from 'drizzle-orm';

const NEWSAPI_KEY = process.env.NEWSAPI_KEY;
const NEWSAPI_BASE = 'https://newsapi.org/v2/everything';

interface NewsApiArticle {
  title: string;
  description: string | null;
  content: string | null;
  url: string;
  urlToImage: string | null;
  publishedAt: string;
  author: string | null;
  source: { name: string };
}

interface NewsApiResponse {
  status: string;
  totalResults: number;
  articles: NewsApiArticle[];
}

async function fetchNewsApi(query: string, pageSize = 20): Promise<NewsApiArticle[]> {
  if (!NEWSAPI_KEY) throw new Error('NEWSAPI_KEY is not set');

  const params = new URLSearchParams({
    q: query,
    language: 'pt',
    sortBy: 'publishedAt',
    pageSize: String(pageSize),
    apiKey: NEWSAPI_KEY,
  });

  const res = await fetch(`${NEWSAPI_BASE}?${params}`, {
    headers: { 'User-Agent': 'fluxnews-bot/1.0' },
    signal: AbortSignal.timeout(15_000),
  });

  if (!res.ok) {
    const body = await res.text();
    throw new Error(`NewsAPI error ${res.status}: ${body}`);
  }

  const data = (await res.json()) as NewsApiResponse;
  if (data.status !== 'ok') throw new Error(`NewsAPI returned status: ${data.status}`);
  return data.articles;
}

async function getExistingSourceUrls(tenant: string): Promise<Set<string>> {
  const existing = await db
    .select({ sourceUrl: posts.sourceUrl })
    .from(posts)
    .where(eq(posts.tenant, tenant));
  return new Set(existing.map((r) => r.sourceUrl).filter(Boolean) as string[]);
}

function slugify(title: string): string {
  return title
    .toLowerCase()
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '')
    .replace(/[^a-z0-9\s-]/g, '')
    .trim()
    .replace(/\s+/g, '-')
    .slice(0, 80);
}

async function processNewsApiSource(source: typeof sources.$inferSelect) {
  const query = source.url; // url field stores the NewsAPI query string
  console.log(`[${source.tenant}] NewsAPI query: "${query}"`);

  let articles: NewsApiArticle[];
  try {
    articles = await fetchNewsApi(query);
  } catch (err) {
    console.error(`  ✗ Failed: ${(err as Error).message}`);
    return { inserted: 0, skipped: 0 };
  }

  const existingUrls = await getExistingSourceUrls(source.tenant);
  let inserted = 0;
  let skipped = 0;

  for (const article of articles) {
    if (!article.url || existingUrls.has(article.url)) {
      skipped++;
      continue;
    }

    // Skip removed articles (NewsAPI placeholder)
    if (article.title === '[Removed]') {
      skipped++;
      continue;
    }

    const slug = `${slugify(article.title)}-${Date.now()}`;
    const excerpt = (article.description ?? '').slice(0, 300).trim();

    try {
      await db.insert(posts).values({
        tenant: source.tenant,
        title: article.title.trim(),
        slug,
        excerpt,
        content: article.content ?? excerpt,
        category: article.source.name,
        author: article.author ?? '',
        sourceUrl: article.url,
        publishedAt: article.publishedAt ? new Date(article.publishedAt) : null,
        status: 'draft',
      });
      existingUrls.add(article.url);
      inserted++;
    } catch (err) {
      console.error(`  ✗ Insert failed: ${(err as Error).message}`);
    }
  }

  await db
    .update(sources)
    .set({ lastFetchedAt: new Date() })
    .where(eq(sources.id, source.id));

  console.log(`  ✓ ${inserted} inserted, ${skipped} skipped`);
  return { inserted, skipped };
}

async function main() {
  console.log('🔄 NewsAPI Fetcher starting…');
  const newsapiSources = await db
    .select()
    .from(sources)
    .where(and(eq(sources.kind, 'newsapi'), eq(sources.active, 1)));

  if (newsapiSources.length === 0) {
    console.log('No active NewsAPI sources found.');
    return;
  }

  let totalInserted = 0;
  let totalSkipped = 0;

  for (const source of newsapiSources) {
    const { inserted, skipped } = await processNewsApiSource(source);
    totalInserted += inserted;
    totalSkipped += skipped;
  }

  console.log(`\n✅ Done. ${totalInserted} articles inserted, ${totalSkipped} skipped.`);
}

main().catch((err) => {
  console.error('Fatal:', err);
  process.exit(1);
});

File: agents/newsapi_fetcher.py
"""
NewsAPI Fetcher Agent
---------------------
Reads active sources (kind='newsapi', url=query string) from the DB,
hits NewsAPI /everything, deduplicates via source_url, inserts drafts.

Usage:
    python agents/newsapi_fetcher.py
    python agents/newsapi_fetcher.py --tenant=cripto --page-size=20
"""

import argparse
import logging
import os
import time

import httpx
from dotenv import load_dotenv

from shared.db import get_db, slugify

load_dotenv()
logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")
log = logging.getLogger(__name__)

NEWSAPI_BASE = "https://newsapi.org/v2/everything"


def fetch_articles(query: str, page_size: int = 20) -> list[dict]:
    key = os.environ.get("NEWSAPI_KEY")
    if not key:
        raise RuntimeError("NEWSAPI_KEY is not set")

    resp = httpx.get(
        NEWSAPI_BASE,
        params={
            "q": query,
            "language": "pt",
            "sortBy": "publishedAt",
            "pageSize": page_size,
            "apiKey": key,
        },
        headers={"User-Agent": "fluxnews-bot/1.0"},
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    if data["status"] != "ok":
        raise RuntimeError(f"NewsAPI error: {data.get('message')}")
    return data["articles"]


def existing_source_urls(conn, tenant: str) -> set[str]:
    rows = conn.execute(
        "SELECT source_url FROM posts WHERE tenant = %s AND source_url IS NOT NULL",
        [tenant],
    ).fetchall()
    return {r["source_url"] for r in rows}


def process_newsapi_source(conn, source: dict, page_size: int) -> tuple[int, int]:
    log.info("[%s] NewsAPI query: '%s'", source["tenant"], source["url"])

    try:
        articles = fetch_articles(source["url"], page_size)
    except Exception as exc:
        log.error("  ✗ Fetch failed: %s", exc)
        return 0, 0

    seen = existing_source_urls(conn, source["tenant"])
    inserted = skipped = 0

    for article in articles:
        url = article.get("url")
        if not url or url in seen or article.get("title") == "[Removed]":
            skipped += 1
            continue

        title = (article["title"] or "").strip()
        excerpt = (article.get("description") or "")[:300].strip()
        content = article.get("content") or excerpt
        author = article.get("author") or ""
        pub_str = article.get("publishedAt")
        pub_date = None
        if pub_str:
            from datetime import datetime, timezone
            pub_date = datetime.fromisoformat(pub_str.replace("Z", "+00:00"))

        slug = f"{slugify(title)}-{int(time.time() * 1000)}"

        try:
            conn.execute(
                """
                INSERT INTO posts
                    (tenant, title, slug, excerpt, content, category, author,
                     source_url, published_at, status)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, 'draft')
                ON CONFLICT DO NOTHING
                """,
                [
                    source["tenant"], title, slug, excerpt, content,
                    article.get("source", {}).get("name", ""), author,
                    url, pub_date,
                ],
            )
            seen.add(url)
            inserted += 1
        except Exception as exc:
            log.error("  ✗ Insert failed for '%s': %s", title[:50], exc)

    conn.execute(
        "UPDATE sources SET last_fetched_at = NOW() WHERE id = %s",
        [source["id"]],
    )
    conn.commit()
    log.info("  ✓ %d inserted, %d skipped", inserted, skipped)
    return inserted, skipped


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--tenant", default=None)
    parser.add_argument("--page-size", type=int, default=20)
    args = parser.parse_args()

    log.info("🔄 NewsAPI Fetcher starting…")
    total_in = total_sk = 0

    with get_db() as conn:
        query = "SELECT * FROM sources WHERE kind = 'newsapi' AND active = 1"
        params: list = []
        if args.tenant:
            query += " AND tenant = %s"
            params.append(args.tenant)
        sources = conn.execute(query, params).fetchall()

        if not sources:
            log.info("No active NewsAPI sources found.")
            return

        for source in sources:
            ins, sk = process_newsapi_source(conn, source, args.page_size)
            total_in += ins
            total_sk += sk

    log.info("✅ Done. %d inserted, %d skipped.", total_in, total_sk)


if __name__ == "__main__":
    main()
